src/candid.ts

```diff
@@ -0,0 +1,167 @@
+        'Subscription-Key': apiKey,
+      },
+    },
+  );
+  logger.debug(`Fetched Candid data for ${ein}: ${response.data.message}`);
+  return response.data;
+};
+
+const getEinBaseFieldId = async (
+  baseUrl: string,
+  token: AccessTokenSet,
+) => {
+  const baseFields = await getBaseFields(baseUrl, token);
+  const einBaseField = baseFields.find(({ shortCode }) => (
+    shortCode === 'organization_tax_id'
+  ));
+  if (einBaseField === undefined) {
+    throw new Error('Could not find base field with short code `organization_tax_id`');
+  }
+  return einBaseField.id;
+};
+
+const isValidEin = (s: string): boolean => (
+  /^\d{2}-?\d{7}$/.test(s)
+);
+
+const getEinsFromPdc = async (baseUrl: string, token: AccessTokenSet) => {
+  const einBaseFieldId = await getEinBaseFieldId(baseUrl, token);
+  const proposals = await getProposals(baseUrl, token);
+
+  const eins = new Set(proposals.entries
+    .flatMap((proposal) => proposal.versions[0]?.fieldValues)
+    .filter(<T>(x: T | undefined): x is T => typeof x !== 'undefined')
+    .filter(({ applicationFormField }) => applicationFormField.baseFieldId === einBaseFieldId)
+    .map(({ value }) => value)
+    .filter(isValidEin));
+  return [...eins];
+};
+
+const candid: CommandModule = {
+  command: 'candid',
+  describe: 'Interact with the Candid Premier API',
+  builder: (y) => (y
+    .option('candid-api-key', {
+      describe: 'Candid Premier API key; get from account management at https://dashboard.candid.org/',
+      demandOption: true,
+      global: true,
+      type: 'string',
+    })
+    .command(
+      'lookup <ein>',
+      'Fetch and display information about an organization by its EIN',
+      {
+        'output-file': {
+          alias: 'write',
+          describe: 'Write organization information to the specified JSON file',
+          normalize: true,
+          type: 'string',
+        },
+      },
+      async (args) => {
+        const result = await getCandidProfile(
+          args.candidApiKey as string,
+          args.ein as string,
+        );
+
+        if (typeof args.outputFile === 'string') {
+          await writeFile(
+            args.outputFile,
+            JSON.stringify(result, null, 2),
+          );
+          logger.info(`Wrote Candid data for ${args.ein as string} to ${args.outputFile}`);
+        } else {
+          logger.info('Candid result', { result });
+        }
+      },
+    )
+    .command(
+      'update <ein>',
+      'Fetch information about an organization by its EIN, and upload to the PDC',
+      oidcOptions,
+      async (args) => {
+        const token = await getToken(
+          args.oidcBaseUrl as string,
+          args.oidcClientId as string,
+          args.oidcClientSecret as string,
+        );
+        const data = await getCandidProfile(args.candidApiKey as string, args.ein as string);
+        await postPlatformProviderData(
+          args.pdcApiBaseUrl as string,
+          token,
+          args.ein as string,
```

Comment:
I just wanted to call out this pattern since I just got bit by it -- this violates type safety, since the actual value is not a string.  This means when doing something like `JSON.stringify(arg.sein)` the result is a number, not a string.

I don't quite know what magic yargs is implicitly doing around arguments (I can't tell where `ein` is actually defined, for instance), but it would be really nice to have all arguments be explicitly typed and provided to us as an explicitly typed value instead of shoehorned to pass type safety checks

Reply:
Oh, that's great feedback! It sounds like I misunderstood how that works.

> I can't tell where `ein` is actually defined

It is defined as an argument to the `update` command via the angle brackets in the command definition:

https://github.com/PhilanthropyDataCommons/data-scripts/blob/9780b998c65f7f07ab87d5fbaf41affc9cb5773a/src/candid.ts#L102

See the docs for [positional arguments](https://github.com/yargs/yargs/blob/main/docs/advanced.md#positional-arguments).

Looks like there's a way to [specify the type of positional arguments](https://github.com/yargs/yargs/blob/main/docs/advanced.md#describing-positional-arguments), which I'd missed - sorry for leaving a trap that bit you, @slifty!

Reply:
> I just wanted to call out this pattern since I just got bit by it -- this violates type safety, since the actual value is not a string. This means when doing something like `JSON.stringify(args.ein)` the result is an encoded number, not a string.

Hm, I wish I had dug into this more deeply while @slifty was around more. I haven't been able to reproduce this.

Positional arguments seem to default to be parsed as `string`s; entering numbers on the command line - and, in particular, numeric expressions like EINs - show up as `string`s for me. Specifying that a positional argument or an option is a `'number'` does cause it to be parsed as such.

The one wrinkle is that if you set an option to be a `string`, and then pass in a `--config` file that has the value as a JSON number, it will get passed in as a number. The reverse is not true: setting the option type to `number`, and passing in a `--config` file with a string-encoded numeric value will be parsed as a number.

---

I also looked more deeply into the original concern I had about needing to add type casts on use, and I think I figured out a solution! We'll need to declare a type for the expected arguments, a la React, and the handler should infer that its `args` is of that type:

```ts
interface ExampleCommandArgs {
  aRequiredString: string;
  anOptionalNumber?: number;
}
const exampleCommand: CommandModule<{}, ExampleCommandArgs> = {
  command: 'example',
  builder: {
    'a-required-string': { type: 'string', demandOption: true },
    'an-optional-number': { type: 'number' },
  },
  handler: (args) => {
    console.log(`It is safe to use string methods ${args.aRequiredString.substring(0)}`);
    if (args.anOptionalNumber) {
      assert(args.anOptionalNumber + 0 !== args.anOptionalNumber + '0');
    }
  },
};
```

This is better than adding `as string` or whatever to every use, although it does not rise to the level of runtime type assertions. This addresses my concern, although I don't know if it addresses @slifty's.